drivers/mrf24j40/mrf24j40_internal.c

```diff
@@ -0,0 +1,244 @@
+    spi_acquire(dev->params.spi);
+    gpio_clear(dev->params.cs_pin);
+    spi_transfer_byte(dev->params.spi, reg1, 0);
+    spi_transfer_byte(dev->params.spi, reg2, &value);
+    spi_transfer_byte(dev->params.spi, 0, &value);
+    gpio_set(dev->params.cs_pin);
+    spi_release(dev->params.spi);
+
+    return (uint8_t)value;
+}
+
+void mrf24j40_reg_write_long(mrf24j40_t *dev, const uint16_t addr, const uint8_t value)
+{
+    uint8_t reg1, reg2;
+
+    reg1 = MRF24J40_LONG_ADDR_TRANS | (addr >> 3);
+    reg2 = (addr << 5) | MRF24J40_ACCESS_WRITE_LNG;
+    spi_acquire(dev->params.spi);
+    gpio_clear(dev->params.cs_pin);
+    spi_transfer_byte(dev->params.spi, reg1, 0);
+    spi_transfer_byte(dev->params.spi, reg2, 0);
+    spi_transfer_byte(dev->params.spi, value, 0);
+    gpio_set(dev->params.cs_pin);
+    spi_release(dev->params.spi);
+}
+
+void mrf24j40_tx_normal_fifo_read(mrf24j40_t *dev, const uint16_t offset, uint8_t *data, const size_t len)
+{
+    uint8_t reg1, reg2;
+
+    reg1 = MRF24J40_LONG_ADDR_TRANS | (offset >> 3);
+    reg2 = (offset << 5) | MRF24J40_ACCESS_READ;
+    spi_acquire(dev->params.spi);
+    gpio_clear(dev->params.cs_pin);
+    spi_transfer_byte(dev->params.spi, reg1, NULL);
+    spi_transfer_byte(dev->params.spi, reg2, (char *)data);
+    spi_transfer_bytes(dev->params.spi, NULL, (char *)data, len);
+    gpio_set(dev->params.cs_pin);
+    spi_release(dev->params.spi);
+}
+
+void mrf24j40_tx_normal_fifo_write(mrf24j40_t *dev,
+                                   const uint16_t offset,
+                                   const uint8_t *data,
+                                   const size_t len)
+{
+    uint16_t addr;
+    uint8_t reg1;
+    uint8_t reg2;
+
+    addr = offset;
+
+    reg1 = MRF24J40_LONG_ADDR_TRANS | (addr >> 3);
+    reg2 = (addr << 5) | MRF24J40_ACCESS_WRITE_LNG;
+
+    spi_acquire(dev->params.spi);
+    gpio_clear(dev->params.cs_pin);
+    spi_transfer_byte(dev->params.spi, reg1, 0);
+    spi_transfer_byte(dev->params.spi, reg2, 0);
+    spi_transfer_bytes(dev->params.spi, (char *)data, NULL, len);
+    gpio_set(dev->params.cs_pin);
+    spi_release(dev->params.spi);
+}
+
+void mrf24j40_rx_fifo_read(mrf24j40_t *dev, const uint16_t offset, uint8_t *data, const size_t len)
+{
+    uint16_t rx_addr;
+
+    rx_addr = MRF24J40_RX_FIFO + offset;
+
+    uint8_t reg1, reg2;
+    reg1 = MRF24J40_LONG_ADDR_TRANS | (rx_addr >> 3);
+    reg2 = (rx_addr << 5) | MRF24J40_ACCESS_READ;
+    spi_acquire(dev->params.spi);
+    gpio_clear(dev->params.cs_pin);
+    spi_transfer_byte(dev->params.spi, reg1, NULL);
+    spi_transfer_byte(dev->params.spi, reg2, (char *)data);
+    spi_transfer_bytes(dev->params.spi, NULL, (char *)data, len);
+    gpio_set(dev->params.cs_pin);
+    spi_release(dev->params.spi);
+}
+
+void mrf24j40_rx_fifo_write(mrf24j40_t *dev, const uint16_t offset, const uint8_t *data, const size_t len)
+{
+    uint16_t i;
+
+    for (i = 0; i < len; i++) {
+        mrf24j40_reg_write_long(dev, i, data[i]);
```

Comment:
No, the mrf24j40 has two types of control registers, short (with 8 bit addresses) and long (16 bit addresses) as the datasheet calls them. The fifo's are in the "long" address space, thus a reg_write_long is needed